### DISCUSSION TITLE
Downloading a file from storage

### DISCUSSION THREAD

**Discussion by nettsite** [filamentphp/filament#2266] [Help] [answered]
Is there a way to download a file from a Table or Page action? I need a modal form to pop up before the file is downloaded.

`
      
     LinkAction::make('bulk_file')

                    ->form([
                        Toggle::make('delete')
                        ->label('Delete file after download?')
                            ->inline(),
                    ])

                    //   Downloads, but does not trigger the form
                    // ->url(fn (Campaign $record): string => route('bulk-file', $record->id), true)
                    
                    // Triggers the form, doesn't download
                    // ->action(function (DownloadBulkPDF $component,Campaign $record, array $data) {
                    //     $component->render($record->id, $data['delete']);
                    // })  

                     // Triggers the form, doesn't download
                    ->action(function (Campaign $record, array $data) {
                        $record->bulk_file($record->id, $data['delete']);
                    });`

This is the method in  my Campaign class:

` 

public function bulk_file($delete = false)
    {
        $action = response($this->bulk_file . 'does not exist. It may have been deleted.', 404);

        if (Storage::disk('output')->exists($this->bulk_file)) {
            $action = Storage::disk('output')->download($this->bulk_file);
        }

        if ($delete) {
            Storage::disk('output')->delete($this->bulk_file);
        }

        Log::debug($action);
        return $action;
    }
`

I have tried these approaches, burt so far no luck, neither in resource table, nor in view page.

Any suggestions?

(I am struggling with the "insert code" thing here, sorry)

**Comment by danharrin** [comment, marked as answer]
You need to return the download from the action, identically to how downloads in Livewire work - https://laravel-livewire.com/docs/2.x/file-downloads

**Reply by nettsite** [reply]
I have actually looked at that documentation, but can't really figure it out - I clearly don't know enough about Livewire.

But thanks for the tip, at least I have a bit of a direction to try a bit harder!

**Reply by danharrin** [reply]
Looking at your code, you're returning the download from the model method, but you're not returning the model method from the action:

```php
 ->action(function (Campaign $record, array $data) {
     return $record->bulk_file($record->id, $data['delete']);
});
```

**Reply by nettsite** [reply]
Here is me licking egg off my face, thank you so much - I have been staring at that for days, clearly in a daze!

Lesson learned (again): when a problem seems insoluble, you are probably over complicating it.
